### DISCUSSION TITLE
Automatic User Migration

### DISCUSSION THREAD

**Discussion by mgraf-dozuki** [zitadel/zitadel#5578] [Ideas]
I am considering Zitadel as a replacement auth solution for an existing B2B multi-tenant SaaS product.  But one big need is the ability to migrate over existing users.

I have seen that users can be imported - as long as we can pass in a bcrypt'ed password.  That does not work for us.  (https://github.com/zitadel/zitadel/discussions/4001)

Other auth solutions have hooks in that allow for an automatic (aka, lazy) migration at the time a user logs in.  The migration is seamless to the user and passwords are migrated securely and automatically without requiring a reset.  

Is this possible (or coming) for Zitadel?

**Comment by fforootd** [comment]
Depending on your current provider you could use SSO in combination with an action to migrate users.

I.e you couple your app to zitadel and auto-register all users with your existing idp.
An action could be used to read additional data of your old idp's api or to manipulate the user.

What also works if you have control over the old system but no OIDC Compliance is to use the [JWT IDP](https://zitadel.com/docs/concepts/structure/jwt_idp) feature to basically forward an existing auth session to zitadel.

**Reply by mgraf-dozuki** [reply]
Thank you Florian for your reply.

There are a couple goals for us as we consider a new IdP.  One is to be able to migrate over existing local users (ie, email/password) without requiring a password reset (thus the desire for an invisible migration to the user).  I have seen this accomplished with a lazy migration in which they enter their password into the IdP, but it recognizes that this is not a current user and hits an endpoint at the old IdP to see if they are valid there and should be migrated over.  After the user is migrated over, their authentication would fully live in the new IdP.

The second goal is to be able to shut down the existing IdP at some point in the future (say 6 months).  Existing users that have not be migrated over already would be considered to be inactive at that point.

Unless I am misunderstanding your suggestions, it seems like both would require the old IdP to live on indefinitely.  Is that correct?

**Reply by fforootd** [reply]
Hm not necessarily. Usually you provide your users with a smooth path for migration where they do not take notice (this either needs the old and new system running in parallel or a lift and shift migration). If you opt for the scenario i mentioned you at one point want to shutdown the old system and provide a way of regaining access for users who are not migrated properly already.

You can create those users in zitadel with their mail verified and a dummy password so that they can trigger a password reset.

A lift and shift migration can be done with https://zitadel.com/docs/apis/admin/admin-service-import-data where you can import passwords (brcypt only ATM) and factors.
